File: loomc-java/src/main/java/loom/compiler/lexer/Lexer.java
package loom.compiler.lexer;

import loom.compiler.token.*;
import loom.compiler.diagnostics.ErrorReporter;

import java.util.ArrayList;
import java.util.List;

public class Lexer {

	private static final List<String> KEYWORDS = Keywords.getKeywords();

	private final String source;
	private final String fileName;
	private final ErrorReporter reporter;
	private final List<Token> tokens = new ArrayList<>();

	private int index = 0;
	private int line = 1;
	private int column = 0;

	public Lexer (String source, String fileName, ErrorReporter reporter) {
		this.source = source;
		this.fileName = fileName;
		this.reporter = reporter;
	}

	public List<Token> tokenize () {
		try {
			while (!isAtEnd()) {
				int tokenStart = index;
				int startLine = line;
				int startColumn = column;

				char c = advance();

				switch (c) {
					case ' ', '\r', '\t' -> {
					} // Skip whitespace
					case '\n' -> {
						line++;
						column = 0;
					}
					case '/' -> {
						if (match('/')) {
							// Single-line comment
							while (!isAtEnd() && peek() != '\n') advance();
						} else if (match('*')) {
							// Block comment
							readBlockComment(startLine, startColumn);
						} else if (match('=')) {
							addToken(TokenType.SLASHEQ, "/=", startLine, startColumn);
						} else {
							addToken(TokenType.SLASH, "/", startLine, startColumn);
						}
					}
					case '"' -> readString(startLine, startColumn);
					case '+' -> {
						if (match('+')) {
							addToken(TokenType.PLUSPLUS, "++", startLine, startColumn);
						} else if (match('=')) {
							addToken(TokenType.PLUSEQ, "+=", startLine, startColumn);
						} else {
							addToken(TokenType.PLUS, "+", startLine, startColumn);
						}
					}
					case '-' -> {
						if (match('-')) {
							addToken(TokenType.MINUSMINUS, "--", startLine, startColumn);
						} else if (match('=')) {
							addToken(TokenType.MINUSEQ, "-=", startLine, startColumn);
						} else if (match('>')) {
							addToken(TokenType.ARROW, "->", startLine, startColumn);
						} else {
							addToken(TokenType.MINUS, "-", startLine, startColumn);
						}
					}
					case '*' -> {
						if (match('=')) {
							addToken(TokenType.STAREQ, "*=", startLine, startColumn);
						} else {
							addToken(TokenType.STAR, "*", startLine, startColumn);
						}
					}
					case '%' -> {
						if (match('=')) {
							addToken(TokenType.MODULOEQ, "%=", startLine, startColumn);
						} else {
							addToken(TokenType.MODULO, "%", startLine, startColumn);
						}
					}
					case '!' ->
							addToken(match('=') ? TokenType.BANGEQ : TokenType.BANG, match('=') ? "!=" : "!", startLine, startColumn);
					case '&' -> {
						if (match('&')) {
							addToken(TokenType.AND, "&&", startLine, startColumn);
						} else {
							addToken(TokenType.BITWISE_AND, "&", startLine, startColumn);
						}
					}
					case '=' -> {
						if (match('=')) {
							addToken(TokenType.EQEQ, "==", startLine, startColumn);
						} else {
							addToken(TokenType.EQ, "=", startLine, startColumn);
						}
					}
					case '<' -> {
						if (match('=')) {
							addToken(TokenType.LTEQ, "<=", startLine, startColumn);
						} else if (match('<')) {
							addToken(TokenType.BITWISE_LSHIFT, "<<", startLine, startColumn);
						} else {
							addToken(TokenType.LT, "<", startLine, startColumn);
						}
					}
					case '>' -> {
						if (match('=')) {
							addToken(TokenType.GTEQ, ">=", startLine, startColumn);
						} else if (match('>')) {
							if (match('>')) {;
								addToken(TokenType.BITWISE_URSHIFT, ">>>", startLine, startColumn);
							} else {
								addToken(TokenType.BITWISE_RSHIFT, ">>", startLine, startColumn);
							}
						} else {
							addToken(TokenType.GT, ">", startLine, startColumn);
						}
					}
					case ';' -> addToken(TokenType.SEMICOLON, ";", startLine, startColumn);
					case ':' -> {
						if (match(':')) {
							addToken(TokenType.DOUBLE_COLON, "::", startLine, startColumn);
						} else {
							addToken(TokenType.COLON, ":", startLine, startColumn);
						}
					}
					case '?' -> addToken(TokenType.QUESTION, "?", startLine, startColumn);
					case '(' -> addToken(TokenType.LPAREN, "(", startLine, startColumn);
					case ')' -> addToken(TokenType.RPAREN, ")", startLine, startColumn);
					case '{' -> addToken(TokenType.LBRACE, "{", startLine, startColumn);
					case '}' -> addToken(TokenType.RBRACE, "}", startLine, startColumn);
					case ',' -> addToken(TokenType.COMMA, ",", startLine, startColumn);
					case '.' -> addToken(TokenType.DOT, ".", startLine, startColumn);
					case '[' -> addToken(TokenType.LBRACKET, "[", startLine, startColumn);
					case ']' -> addToken(TokenType.RBRACKET, "]", startLine, startColumn);
					case '|' -> {
						if (match('|')) {
							addToken(TokenType.OR, "||", startLine, startColumn);
						} else {
							addToken(TokenType.BITWISE_OR, "|", startLine, startColumn);
						}
					}
					case '^' -> addToken(TokenType.BITWISE_XOR, "^", startLine, startColumn);
					case '~' -> addToken(TokenType.BITWISE_NOT, "~", startLine, startColumn);
					default -> {
						if (Character.isDigit(c)) {
							readNumber(c, startLine, startColumn);
						} else if (isIdentifierStart(c)) {
							readIdentifier(c, startLine, startColumn);
						} else {
							// Emit an UNKNOWN token
							String badChar = String.valueOf(c);
							TokenSpan span = TokenSpan.singleLine(startLine, startColumn, startColumn + 1, fileName);
							reporter.reportError(span, "Unexpected character: '" + badChar + "'", source);
							tokens.add(new Token(TokenType.UNKNOWN, badChar, span));
						}
					}
				}
			}
		} catch (Exception e) {
			TokenSpan span = TokenSpan.singleLine(line, column, column + 1, fileName);
			reporter.reportError(span, "Internal lexer failure: " + e.getMessage(), source);
		}

		// Add EOF token - ensure it doesn't extend beyond source code
		int eofColumn = Math.min(column, source.length());
		tokens.add(new Token(TokenType.EOF, "", TokenSpan.singleLine(line, eofColumn, eofColumn, fileName)));
		return tokens;
	}

	private boolean isIdentifierStart(char c) {
		return Character.isLetter(c) || c == '_';
	}

	private boolean isIdentifierPart(char c) {
		return Character.isLetterOrDigit(c) || c == '_';
	}

	private boolean isAtEnd () {
		return index >= source.length();
	}

	private char peek () {
		return isAtEnd() ? '\0' : source.charAt(index);
	}

	private char advance () {
		if (isAtEnd()) return '\0';
		char c = source.charAt(index++);
		column++;
		return c;
	}

	private boolean match (char expected) {
		if (isAtEnd() || source.charAt(index) != expected) return false;
		index++;
		column++;
		return true;
	}

	private void addToken (TokenType type, String value, int startLine, int startCol) {
		int endCol = startCol + value.length();
		TokenSpan span = TokenSpan.singleLine(startLine, startCol, endCol, fileName);
		tokens.add(new Token(type, value, span));
	}

	private void readString (int startLine, int startCol) {
		StringBuilder builder = new StringBuilder();
		while (!isAtEnd() && peek() != '"') {
			if (peek() == '\n') {
				line++;
				column = 0;
			}
			builder.append(advance());
		}
		if (isAtEnd()) {
			reporter.reportError(TokenSpan.singleLine(startLine, startCol, column, fileName), "Unterminated string", source);
			return;
		}
		advance(); // closing quote
		addToken(TokenType.STRING, builder.toString(), startLine, startCol);
	}

	private void readNumber (char first, int startLine, int startCol) {
		StringBuilder builder = new StringBuilder();
		builder.append(first);

		while (Character.isDigit(peek())) builder.append(advance());

		if (peek() == '.' && Character.isDigit(peekNext())) {
			do builder.append(advance());
			while (Character.isDigit(peek()));
		}

		addToken(TokenType.NUMBER, builder.toString(), startLine, startCol);
	}

	private char peekNext () {
		return (index + 1 < source.length()) ? source.charAt(index + 1) : '\0';
	}

	private void readIdentifier(char first, int startLine, int startCol) {
		StringBuilder builder = new StringBuilder();
		builder.append(first);

		while (!isAtEnd() && isIdentifierPart(peek())) {
			builder.append(advance());
		}

		String value = builder.toString();
		TokenType type = KEYWORDS.contains(value) ? TokenType.KEYWORD : TokenType.IDENTIFIER;
		addToken(type, value, startLine, startCol);
	}

	private void readBlockComment(int startLine, int startCol) {
		// Skip the opening /* and start reading the comment content
		while (!isAtEnd()) {
			if (peek() == '*' && peekNext() == '/') {
				// Found the closing */
				advance(); // consume *
				advance(); // consume /
				return;
			}
			
			char c = advance();
			if (c == '\n') {
				line++;
				column = 0;
			}
		}
		
		// If we reach here, the block comment was not properly closed
		reporter.reportError(TokenSpan.singleLine(startLine, startCol, column, fileName), 
			"Unterminated block comment", source);
	}
}

File: loomc/lib/lexer.ml
(** Lexer for the Loom language *)

(** Lexer state *)
type lexer_state = {
  source : string;
  file_name : string;
  mutable tokens : Token.t list;
  mutable index : int;
  mutable line : int;
  mutable column : int;
}

(** Create a new lexer *)
let create source file_name =
  { source; file_name; tokens = []; index = 0; line = 1; column = 0 }

(** Check if we're at the end of input *)
let is_at_end state = state.index >= String.length state.source

(** Peek at the current character *)
let peek state =
  if is_at_end state then '\000'
  else String.get state.source state.index

(** Peek at the next character *)
let peek_next state =
  if state.index + 1 >= String.length state.source then '\000'
  else String.get state.source (state.index + 1)

(** Advance to the next character *)
let advance state =
  if is_at_end state then '\000'
  else (
    let c = String.get state.source state.index in
    state.index <- state.index + 1;
    state.column <- state.column + 1;
    c
  )

(** Match and consume a character if it matches *)
let match_char state expected =
  if is_at_end state then false
  else if String.get state.source state.index = expected then (
    let _ = advance state in
    true
  ) else false

(** Add a token to the lexer state *)
let add_token state token_type value start_line start_col =
  let span = Token_span.single_line start_line start_col state.column (Some state.file_name) in
  let token = Token.make token_type value span in
  state.tokens <- token :: state.tokens

(** Check if a character can start an identifier *)
let is_identifier_start c =
  (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c = '_'

(** Check if a character can be part of an identifier *)
let is_identifier_part c =
  (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || 
  (c >= '0' && c <= '9') || c = '_'

(** Read a string literal *)
let read_string state start_line start_col =
  let buf = Buffer.create 16 in
  let rec loop () =
    if is_at_end state then (
      (* Unterminated string - we'll handle this as an error later *)
      ()
    ) else (
      let c = advance state in
      match c with
      | '"' -> ()
      | '\\' -> (
          if is_at_end state then (
            (* Unterminated escape sequence *)
            ()
          ) else (
            let next = advance state in
            match next with
            | 'n' -> Buffer.add_char buf '\n'
            | 't' -> Buffer.add_char buf '\t'
            | 'r' -> Buffer.add_char buf '\r'
            | '\\' -> Buffer.add_char buf '\\'
            | '"' -> Buffer.add_char buf '"'
            | _ -> Buffer.add_char buf next
          );
          loop ()
        )
      | _ -> Buffer.add_char buf c; loop ()
    )
  in
  loop ();
  let value = Buffer.contents buf in
  add_token state Token_types.String value start_line start_col

(** Read a number literal *)
let read_number state first start_line start_col =
  let buf = Buffer.create 16 in
  Buffer.add_char buf first;
  
  let rec loop () =
    if is_at_end state then ()
    else (
      let c = peek state in
      if (c >= '0' && c <= '9') || c = '.' || c = 'e' || c = 'E' || c = '+' || c = '-' then (
        let _ = advance state in
        Buffer.add_char buf c;
        loop ()
      )
    )
  in
  loop ();
  
  let value = Buffer.contents buf in
  add_token state Token_types.Number value start_line start_col

(** Read an identifier or keyword *)
let read_identifier state first start_line start_col =
  let buf = Buffer.create 16 in
  Buffer.add_char buf first;
  
  let rec loop () =
    if is_at_end state then ()
    else (
      let c = peek state in
      if is_identifier_part c then (
        let _ = advance state in
        Buffer.add_char buf c;
        loop ()
      )
    )
  in
  loop ();
  
  let value = Buffer.contents buf in
  let token_type = if Keywords.is_keyword value then Token_types.Keyword else Token_types.Identifier in
  add_token state token_type value start_line start_col

(** Read a block comment *)
let read_block_comment state start_line start_col =
  let rec loop () =
    if is_at_end state then (
      (* Unterminated comment - we'll handle this as an error later *)
      ()
    ) else (
      let c = advance state in
      match c with
      | '*' -> (
          if not (is_at_end state) && peek state = '/' then (
            let _ = advance state in
            ()
          ) else loop ()
        )
      | '\n' -> state.line <- state.line + 1; state.column <- 0; loop ()
      | _ -> loop ()
    )
  in
  loop ();
  
  let span = Token_span.make start_line start_col state.line state.column (Some state.file_name) in
  let token = Token.make Token_types.Comment "" span in
  state.tokens <- token :: state.tokens

(** Tokenize the source code *)
let tokenize state =
  while not (is_at_end state) do
    let start_line = state.line in
    let start_col = state.column in
    
    let c = advance state in
    
    match c with
    | ' ' | '\r' | '\t' -> () (* Skip whitespace *)
    | '\n' -> state.line <- state.line + 1; state.column <- 0
    | '/' -> (
        if match_char state '/' then (
          (* Single-line comment *)
          while not (is_at_end state) && peek state != '\n' do
            let _ = advance state in ()
          done
        ) else if match_char state '*' then (
          (* Block comment *)
          read_block_comment state start_line start_col
        ) else if match_char state '=' then (
          add_token state Token_types.Slasheq "/=" start_line start_col
        ) else (
          add_token state Token_types.Slash "/" start_line start_col
        )
      )
    | '"' -> read_string state start_line start_col
    | '+' -> (
        if match_char state '+' then (
          add_token state Token_types.Plusplus "++" start_line start_col
        ) else if match_char state '=' then (
          add_token state Token_types.Pluseq "+=" start_line start_col
        ) else (
          add_token state Token_types.Plus "+" start_line start_col
        )
      )
    | '-' -> (
        if match_char state '-' then (
          add_token state Token_types.Minusminus "--" start_line start_col
        ) else if match_char state '=' then (
          add_token state Token_types.Minuseq "-=" start_line start_col
        ) else if match_char state '>' then (
          add_token state Token_types.Arrow "->" start_line start_col
        ) else (
          add_token state Token_types.Minus "-" start_line start_col
        )
      )
    | '*' -> (
        if match_char state '=' then (
          add_token state Token_types.Stareq "*=" start_line start_col
        ) else (
          add_token state Token_types.Star "*" start_line start_col
        )
      )
    | '%' -> (
        if match_char state '=' then (
          add_token state Token_types.Moduloeq "%=" start_line start_col
        ) else (
          add_token state Token_types.Modulo "%" start_line start_col
        )
      )
    | '!' -> (
        if match_char state '=' then (
          add_token state Token_types.Bangeq "!=" start_line start_col
        ) else (
          add_token state Token_types.Bang "!" start_line start_col
        )
      )
    | '&' -> (
        if match_char state '&' then (
          add_token state Token_types.And "&&" start_line start_col
        ) else (
          add_token state Token_types.Bitwise_and "&" start_line start_col
        )
      )
    | '|' -> (
        if match_char state '|' then (
          add_token state Token_types.Or "||" start_line start_col
        ) else (
          add_token state Token_types.Bitwise_or "|" start_line start_col
        )
      )
    | '^' -> add_token state Token_types.Bitwise_xor "^" start_line start_col
    | '~' -> add_token state Token_types.Bitwise_not "~" start_line start_col
    | '<' -> (
        if match_char state '<' then (
          add_token state Token_types.Bitwise_lshift "<<" start_line start_col
        ) else if match_char state '=' then (
          add_token state Token_types.Lteq "<=" start_line start_col
        ) else (
          add_token state Token_types.Lt "<" start_line start_col
        )
      )
    | '>' -> (
        if match_char state '>' then (
          if match_char state '>' then (
            add_token state Token_types.Bitwise_urshift ">>>" start_line start_col
          ) else (
            add_token state Token_types.Bitwise_rshift ">>" start_line start_col
          )
        ) else if match_char state '=' then (
          add_token state Token_types.Gteq ">=" start_line start_col
        ) else (
          add_token state Token_types.Gt ">" start_line start_col
        )
      )
    | '=' -> (
        if match_char state '=' then (
          add_token state Token_types.Eqeq "==" start_line start_col
        ) else (
          add_token state Token_types.Eq "=" start_line start_col
        )
      )
    | '(' -> add_token state Token_types.Lparen "(" start_line start_col
    | ')' -> add_token state Token_types.Rparen ")" start_line start_col
    | '{' -> add_token state Token_types.Lbrace "{" start_line start_col
    | '}' -> add_token state Token_types.Rbrace "}" start_line start_col
    | '[' -> add_token state Token_types.Lbracket "[" start_line start_col
    | ']' -> add_token state Token_types.Rbracket "]" start_line start_col
    | ',' -> add_token state Token_types.Comma "," start_line start_col
    | ';' -> add_token state Token_types.Semicolon ";" start_line start_col
    | ':' -> (
        if match_char state ':' then (
          add_token state Token_types.Double_colon "::" start_line start_col
        ) else (
          add_token state Token_types.Colon ":" start_line start_col
        )
      )
    | '?' -> add_token state Token_types.Question "?" start_line start_col
    | '.' -> add_token state Token_types.Dot "." start_line start_col
    | c when is_identifier_start c -> read_identifier state c start_line start_col
    | c when (c >= '0' && c <= '9') -> read_number state c start_line start_col
    | c -> add_token state Token_types.Unknown (String.make 1 c) start_line start_col
  done;
  
  (* Add EOF token *)
  let eof_span = Token_span.single_line state.line state.column state.column (Some state.file_name) in
  let eof_token = Token.make Token_types.Eof "" eof_span in
  state.tokens <- eof_token :: state.tokens;
  
  (* Return tokens in reverse order (they were added in reverse) *)
  List.rev state.tokens

(** Tokenize a string *)
let tokenize_string source file_name =
  let state = create source file_name in
  tokenize state 